server/analyzer/comparison_casts.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/dolthub/doltgresql/server/functions/framework"
+	pgtypes "github.com/dolthub/doltgresql/server/types"
+)
+
+// ComparisonCasts handles implicit conversions for comparisons. We cannot yet implement the comparison operators as GMS
+// relies on specific expressions some cases. Eventually, we'll need to decouple the dependency, so that we may use the
+// appropriate functions. This is technically incorrect, but should let us make progress until then.
+func ComparisonCasts(ctx *sql.Context, a *analyzer.Analyzer, node sql.Node, scope *plan.Scope, selector analyzer.RuleSelector) (sql.Node, transform.TreeIdentity, error) {
+	if disjointedNode, ok := node.(plan.DisjointedChildrenNode); ok {
+		return handleDisjointedNodes(ctx, a, disjointedNode, scope, selector, ComparisonCasts)
+	}
+	return transform.NodeExprsWithOpaque(node, func(expr sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
+		switch comparison := expr.(type) {
+		case *expression.Equals:
+			left, right, ok, err := comparisonCasts(comparison.LeftChild, comparison.RightChild)
+			if err != nil {
+				return nil, transform.NewTree, err
+			}
+			if !ok {
+				return comparison, transform.SameTree, nil
+			}
+			return expression.NewEquals(left, right), transform.NewTree, nil
+		case *expression.NullSafeEquals:
+			left, right, ok, err := comparisonCasts(comparison.LeftChild, comparison.RightChild)
+			if err != nil {
+				return nil, transform.NewTree, err
+			}
+			if !ok {
+				return comparison, transform.SameTree, nil
+			}
+			return expression.NewNullSafeEquals(left, right), transform.NewTree, nil
+		case *expression.GreaterThan:
+			left, right, ok, err := comparisonCasts(comparison.LeftChild, comparison.RightChild)
+			if err != nil {
+				return nil, transform.NewTree, err
+			}
+			if !ok {
+				return comparison, transform.SameTree, nil
+			}
+			return expression.NewGreaterThan(left, right), transform.NewTree, nil
+		case *expression.GreaterThanOrEqual:
+			left, right, ok, err := comparisonCasts(comparison.LeftChild, comparison.RightChild)
+			if err != nil {
+				return nil, transform.NewTree, err
+			}
+			if !ok {
+				return comparison, transform.SameTree, nil
+			}
+			return expression.NewGreaterThanOrEqual(left, right), transform.NewTree, nil
+		case *expression.LessThan:
+			left, right, ok, err := comparisonCasts(comparison.LeftChild, comparison.RightChild)
+			if err != nil {
+				return nil, transform.NewTree, err
+			}
+			if !ok {
+				return comparison, transform.SameTree, nil
+			}
+			return expression.NewLessThan(left, right), transform.NewTree, nil
+		case *expression.LessThanOrEqual:
+			left, right, ok, err := comparisonCasts(comparison.LeftChild, comparison.RightChild)
+			if err != nil {
+				return nil, transform.NewTree, err
+			}
+			if !ok {
+				return comparison, transform.SameTree, nil
+			}
+			return expression.NewLessThanOrEqual(left, right), transform.NewTree, nil
+		default:
+			return expr, transform.SameTree, nil
```

Comment:
More appropriate to error here? What's the consequence of an unhandled comparison? Should it here